runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java

```diff
@@ -0,0 +1,270 @@
+   * {@link StateRequestHandler}.
+   *
+   * <p>The {@link MultimapSideInputHandlerFactory} is required to handle all multimap side inputs
+   * contained within the {@link ExecutableProcessBundleDescriptor}. See
+   * {@link ExecutableProcessBundleDescriptor#getMultimapSideInputSpecs} for the set of multimap
+   * side inputs that are contained.
+   */
+  public static StateRequestHandler forMultimapSideInputHandlerFactory(
+      ExecutableProcessBundleDescriptor processBundleDescriptor,
+      MultimapSideInputHandlerFactory multimapSideInputHandlerFactory) {
+    return new StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter(
+        processBundleDescriptor, multimapSideInputHandlerFactory);
+  }
+
+  /**
+   * An adapter which converts {@link MultimapSideInputHandlerFactory} to
+   * {@link StateRequestHandler}.
+   */
+  static class StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter
+      implements StateRequestHandler {
+
+    private final ExecutableProcessBundleDescriptor processBundleDescriptor;
+    private final MultimapSideInputHandlerFactory multimapSideInputHandlerFactory;
+    private final ConcurrentHashMap<MultimapSideInputSpec, MultimapSideInputHandler> cache;
+
+    StateRequestHandlerToMultimapSideInputHandlerFactoryAdapter(
+        ExecutableProcessBundleDescriptor processBundleDescriptor,
+        MultimapSideInputHandlerFactory multimapSideInputHandlerFactory) {
+      this.processBundleDescriptor = processBundleDescriptor;
+      this.multimapSideInputHandlerFactory = multimapSideInputHandlerFactory;
+      this.cache = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public CompletionStage<StateResponse.Builder> handle(
+        StateRequest request) throws Exception {
+      try {
+        checkState(TypeCase.MULTIMAP_SIDE_INPUT.equals(request.getStateKey().getTypeCase()),
+            "Unsupported %s type %s, expected %s",
+            StateRequest.class.getSimpleName(),
+            request.getStateKey().getTypeCase(),
+            TypeCase.MULTIMAP_SIDE_INPUT);
+
+        StateKey.MultimapSideInput stateKey = request.getStateKey().getMultimapSideInput();
+        MultimapSideInputSpec<?, ?, ?> sideInputReferenceSpec =
+            processBundleDescriptor.getMultimapSideInputSpecs()
+                .get(stateKey.getPtransformId())
+                .get(stateKey.getSideInputId());
+        MultimapSideInputHandler<?, ?, ?> handler = cache.computeIfAbsent(
+            sideInputReferenceSpec,
+            this::createHandler);
+
+        switch (request.getRequestCase()) {
+          case GET:
+            return handleGetRequest(request, handler);
+          case APPEND:
+          case CLEAR:
+          default:
+            throw new Exception(String.format(
+                "Unsupported request type %s for side input.", request.getRequestCase()));
+        }
+      } catch (Exception e) {
+        CompletableFuture f = new CompletableFuture();
+        f.completeExceptionally(e);
```

Comment:
It's unfortunate that there's no static constructor for exceptional futures. 😞

Reply:
Unfortunate